File: juce_backend/src/audio/mixing/mixing_console.cpp
/*
 * mixing_console.cpp
 * White Room Audio Engine - Mixing Console Implementation
 */

#include "mixing_console.h"
#include <algorithm>
#include <cmath>

namespace white_room {
namespace audio {

MixingConsoleProcessor::MixingConsoleProcessor() {
    // Create master bus
    masterBus = std::make_unique<ChannelStrip>();
    masterBus->id = 0;
    masterBus->name = "Master";
    masterBus->type = "master";
    masterBus->volume = 0.8f;
    masterBus->pan = 0.0f;
}

MixingConsoleProcessor::~MixingConsoleProcessor() {
    channels.clear();
    masterBus.reset();
}

// ========== Channel Management ==========

void MixingConsoleProcessor::addChannel(std::unique_ptr<ChannelStrip> channel) {
    channels.push_back(std::move(channel));
}

void MixingConsoleProcessor::removeChannel(int id) {
    auto it = std::remove_if(channels.begin(), channels.end(),
        [id](const auto& ch) { return ch->id == id; });
    channels.erase(it, channels.end());
}

ChannelStrip* MixingConsoleProcessor::getChannel(int id) {
    if (id == 0) return masterBus.get();

    auto it = std::find_if(channels.begin(), channels.end(),
        [id](const auto& ch) { return ch->id == id; });
    return (it != channels.end()) ? it->get() : nullptr;
}

std::vector<ChannelStrip*> MixingConsoleProcessor::getAllChannels() {
    std::vector<ChannelStrip*> allChannels;
    for (auto& channel : channels) {
        allChannels.push_back(channel.get());
    }
    allChannels.push_back(masterBus.get());
    return allChannels;
}

ChannelStrip* MixingConsoleProcessor::getMasterBus() {
    return masterBus.get();
}

// ========== Audio Processing ==========

void MixingConsoleProcessor::prepareToPlay(double sampleRate, int samplesPerBlock) {
    currentSampleRate = sampleRate;

    // Prepare mix buffer (stereo)
    mixBuffer.setSize(2, samplesPerBlock);
    channelBuffer.setSize(2, samplesPerBlock);
}

void MixingConsoleProcessor::reset() {
    mixBuffer.clear();
    channelBuffer.clear();

    for (auto& channel : channels) {
        channel->levelL = -60.0f;
        channel->levelR = -60.0f;
        channel->peakL = -60.0f;
        channel->peakR = -60.0f;
    }

    masterBus->levelL = -60.0f;
    masterBus->levelR = -60.0f;
    masterBus->peakL = -60.0f;
    masterBus->peakR = -60.0f;
}

void MixingConsoleProcessor::processBlock(juce::AudioBuffer<float>& buffer) {
    const int numSamples = buffer.getNumSamples();
    const int numChannels = buffer.getNumChannels();

    // Clear mix buffer
    mixBuffer.clear();

    // Process each channel
    for (auto& channel : channels) {
        if (!shouldPlayChannel(*channel)) {
            // Muted channel - clear levels
            channel->levelL = -60.0f;
            channel->levelR = -60.0f;
            continue;
        }

        // Copy channel data to temp buffer
        channelBuffer.setSize(buffer.getNumChannels(), numSamples, false, false, true);
        channelBuffer.copyFrom(0, 0, buffer, 0, 0, numSamples);

        if (buffer.getNumChannels() >= 2) {
            channelBuffer.copyFrom(1, 0, buffer, 1, 0, numSamples);
        }

        // Apply volume and pan
        applyVolumePan(*channel, channelBuffer);

        // Add to mix buffer
        mixBuffer.addFrom(0, 0, channelBuffer, 0, 0, numSamples);
        mixBuffer.addFrom(1, 0, channelBuffer, 1, 0, numSamples);

        // Update metering
        updateMetering(*channel, channelBuffer);
    }

    // Apply master bus processing
    applyVolumePan(*masterBus, mixBuffer);

    // Copy mix buffer to output
    buffer.copyFrom(0, 0, mixBuffer, 0, 0, numSamples);
    if (numChannels >= 2) {
        buffer.copyFrom(1, 0, mixBuffer, 1, 0, numSamples);
    }

    // Update master metering
    updateMetering(*masterBus, mixBuffer);
}

// ========== Level Controls ==========

void MixingConsoleProcessor::setVolume(int channelId, float volume) {
    ChannelStrip* channel = getChannel(channelId);
    if (channel) {
        channel->volume = juce::jlimit(0.0f, 1.0f, volume);
    }
}

void MixingConsoleProcessor::setPan(int channelId, float pan) {
    ChannelStrip* channel = getChannel(channelId);
    if (channel) {
        channel->pan = juce::jlimit(-1.0f, 1.0f, pan);
    }
}

void MixingConsoleProcessor::setMute(int channelId, bool muted) {
    ChannelStrip* channel = getChannel(channelId);
    if (channel) {
        channel->isMuted = muted;
    }
}

void MixingConsoleProcessor::setSolo(int channelId, bool solo) {
    ChannelStrip* channel = getChannel(channelId);
    if (channel && channel->type != "master") {
        channel->isSolo = solo;

        // Mute all non-soloed channels if any are soloed
        bool hasSolo = hasSoloedChannels();
        for (auto& ch : channels) {
            if (ch->id != channelId && ch->type != "master") {
                ch->isMuted = hasSolo && !ch->isSolo;
            }
        }
    }
}

// ========== Metering ==========

float MixingConsoleProcessor::getLevelL(int channelId) const {
    const ChannelStrip* channel = const_cast<MixingConsoleProcessor*>(this)->getChannel(channelId);
    return channel ? channel->levelL : -60.0f;
}

float MixingConsoleProcessor::getLevelR(int channelId) const {
    const ChannelStrip* channel = const_cast<MixingConsoleProcessor*>(this)->getChannel(channelId);
    return channel ? channel->levelR : -60.0f;
}

float MixingConsoleProcessor::getPeakL(int channelId) const {
    const ChannelStrip* channel = const_cast<MixingConsoleProcessor*>(this)->getChannel(channelId);
    return channel ? channel->peakL : -60.0f;
}

float MixingConsoleProcessor::getPeakR(int channelId) const {
    const ChannelStrip* channel = const_cast<MixingConsoleProcessor*>(this)->getChannel(channelId);
    return channel ? channel->peakR : -60.0f;
}

std::map<int, std::pair<float, float>> MixingConsoleProcessor::getAllMeterData() const {
    std::map<int, std::pair<float, float>> meterData;

    for (const auto& channel : channels) {
        meterData[channel->id] = {channel->levelL, channel->levelR};
    }

    meterData[0] = {masterBus->levelL, masterBus->levelR};

    return meterData;
}

// ========== Routing ==========

void MixingConsoleProcessor::setOutputBus(int channelId, const juce::String& bus) {
    ChannelStrip* channel = getChannel(channelId);
    if (channel && channel->type != "master") {
        channel->outputBus = bus;
    }
}

// ========== Internal Processing ==========

void MixingConsoleProcessor::applyVolumePan(ChannelStrip& channel, juce::AudioBuffer<float>& buffer) {
    const int numSamples = buffer.getNumSamples();

    // Apply pan law (-3dB at center)
    float panL = 1.0f;
    float panR = 1.0f;

    if (channel.pan < 0.0f) {
        // Pan left
        panL = 1.0f;
        panR = juce::jmin(1.0f, 1.0f + channel.pan);
    } else if (channel.pan > 0.0f) {
        // Pan right
        panL = juce::jmin(1.0f, 1.0f - channel.pan);
        panR = 1.0f;
    }

    // Apply volume and pan
    float volL = channel.volume * panL;
    float volR = channel.volume * panR;

    buffer.applyGain(0, 0, numSamples, volL);
    if (buffer.getNumChannels() >= 2) {
        buffer.applyGain(1, 0, numSamples, volR);
    }
}

bool MixingConsoleProcessor::shouldPlayChannel(const ChannelStrip& channel) const {
    if (channel.isMuted) return false;

    // If any channel is soloed, only play soloed channels
    if (hasSoloedChannels()) {
        return channel.isSolo;
    }

    return true;
}

void MixingConsoleProcessor::updateMetering(ChannelStrip& channel, const juce::AudioBuffer<float>& buffer) {
    const int numSamples = buffer.getNumSamples();

    // Find RMS and peak levels
    float rmsL = 0.0f;
    float rmsR = 0.0f;
    float peakL = 0.0f;
    float peakR = 0.0f;

    for (int i = 0; i < numSamples; ++i) {
        float sampleL = std::abs(buffer.getSample(0, i));
        peakL = juce::jmax(peakL, sampleL);
        rmsL += sampleL * sampleL;

        if (buffer.getNumChannels() >= 2) {
            float sampleR = std::abs(buffer.getSample(1, i));
            peakR = juce::jmax(peakR, sampleR);
            rmsR += sampleR * sampleR;
        }
    }

    rmsL = std::sqrt(rmsL / numSamples);
    rmsR = std::sqrt(rmsR / numSamples);

    // Convert to dB
    float levelL_new = linearToDecibels(rmsL);
    float levelR_new = linearToDecibels(rmsR);
    float peakL_new = linearToDecibels(peakL);
    float peakR_new = linearToDecibels(peakR);

    // Smooth metering
    float smoothing = 0.2f;
    channel.levelL = channel.levelL * (1.0f - smoothing) + levelL_new * smoothing;
    channel.levelR = channel.levelR * (1.0f - smoothing) + levelR_new * smoothing;

    // Peak hold (instant attack, slow decay)
    channel.peakL = juce::jmax(channel.peakL - 0.5f, peakL_new);
    channel.peakR = juce::jmax(channel.peakR - 0.5f, peakR_new);
}

float MixingConsoleProcessor::linearToDecibels(float linear) {
    if (linear < 1e-6f) return -60.0f;
    return 20.0f * std::log10(linear);
}

float MixingConsoleProcessor::decibelsToLinear(float db) {
    if (db <= -60.0f) return 0.0f;
    return std::pow(10.0f, db / 20.0f);
}

bool MixingConsoleProcessor::hasSoloedChannels() const {
    return std::any_of(channels.begin(), channels.end(),
        [](const auto& ch) { return ch->isSolo; });
}

} // namespace audio
} // namespace white_room

File: sdk/src/models/MixingConsole.ts
/**
 * Mixing Console Data Models
 *
 * Professional mixing console with channel strips, effects, and automation
 */

/**
 * Insert slot for channel strip effects (pre-fader)
 */
export interface InsertSlot {
  id: string
  enabled: boolean
  plugin?: string // Plugin ID if virtual instrument
  effect?: string // Effect type (e.g., 'compressor', 'eq', 'reverb')
  parameters: Record<string, number>
}

/**
 * Send effect (post-fader to bus)
 */
export interface Send {
  id: string
  bus: string // Target bus ID
  amount: number // 0-1 (send level)
  prePost: 'pre' | 'post' // Pre-fader or post-fader send
}

/**
 * Metering data for stereo channels
 */
export interface MeterData {
  levelL: number // Current level in dB (-60 to 0)
  levelR: number
  peakL: number // Peak hold level
  peakR: number
  rmsL: number // RMS level
  rmsR: number
}

/**
 * Channel strip - individual mixing channel
 */
export interface ChannelStrip {
  id: string
  name: string
  type: 'audio' | 'midi' | 'bus' | 'master'

  // Level controls
  volume: number // 0-1 (linear)
  pan: number // -1 to 1 (left to right)

  // Mute/Solo
  isMuted: boolean
  isSolo: boolean

  // Metering
  levelL: number // Current level (dB)
  levelR: number
  peakL: number
  peakR: number

  // Effects inserts (pre-fader)
  inserts: InsertSlot[]

  // Effects sends (post-fader)
  sends: Send[]

  // Routing
  outputBus: string // Where this channel routes to
}

/**
 * Automation point for parameter automation
 */
export interface AutomationPoint {
  time: number // Time in seconds
  value: number // Parameter value
  curve?: 'linear' | 'exponential' | 'logarithmic'
}

/**
 * Automation curve for a single parameter
 */
export interface AutomationCurve {
  parameterId: string // e.g., 'channel_1.volume'
  points: AutomationPoint[]
}

/**
 * Mixing console state and management
 */
export class MixingConsole {
  private channels: Map<string, ChannelStrip>
  private masterBus: ChannelStrip
  private automation: Map<string, AutomationCurve>
  private isPlayingAutomation: boolean = false

  constructor() {
    this.channels = new Map()
    this.automation = new Map()
    this.masterBus = this.createMasterBus()
  }

  /**
   * Create master bus channel
   */
  private createMasterBus(): ChannelStrip {
    return {
      id: 'master',
      name: 'Master',
      type: 'master',
      volume: 0.8,
      pan: 0,
      isMuted: false,
      isSolo: false,
      levelL: -60,
      levelR: -60,
      peakL: -60,
      peakR: -60,
      inserts: [],
      sends: [],
      outputBus: ''
    }
  }

  // ========== Channel Management ==========

  /**
   * Add a new channel to the console
   */
  addChannel(channel: ChannelStrip): void {
    this.channels.set(channel.id, channel)
  }

  /**
   * Remove a channel from the console
   */
  removeChannel(id: string): void {
    this.channels.delete(id)
  }

  /**
   * Get a specific channel by ID
   */
  getChannel(id: string): ChannelStrip | undefined {
    if (id === 'master') return this.masterBus
    return this.channels.get(id)
  }

  /**
   * Get all channels including master
   */
  getAllChannels(): ChannelStrip[] {
    return [
      ...Array.from(this.channels.values()),
      this.masterBus
    ]
  }

  /**
   * Get all non-master channels
   */
  getMixChannels(): ChannelStrip[] {
    return Array.from(this.channels.values())
  }

  /**
   * Get master bus
   */
  getMasterBus(): ChannelStrip {
    return this.masterBus
  }

  // ========== Level Controls ==========

  /**
   * Set channel volume
   */
  setVolume(id: string, volume: number): void {
    const channel = this.getChannel(id)
    if (channel) {
      channel.volume = Math.max(0, Math.min(1, volume))
      this.recordAutomation(`${id}.volume`, channel.volume)
    }
  }

  /**
   * Set channel pan
   */
  setPan(id: string, pan: number): void {
    const channel = this.getChannel(id)
    if (channel) {
      channel.pan = Math.max(-1, Math.min(1, pan))
      this.recordAutomation(`${id}.pan`, channel.pan)
    }
  }

  /**
   * Set channel mute state
   */
  setMute(id: string, muted: boolean): void {
    const channel = this.getChannel(id)
    if (channel) {
      // If soloing any channel, mute all non-soloed channels
      if (muted === false && this.hasSoloedChannels()) {
        return // Don't unmute if other channels are soloed
      }
      channel.isMuted = muted
      this.recordAutomation(`${id}.mute`, muted ? 1 : 0)
    }
  }

  /**
   * Set channel solo state
   */
  setSolo(id: string, solo: boolean): void {
    const channel = this.getChannel(id)
    if (channel && channel.type !== 'master') {
      channel.isSolo = solo

      // Mute all non-soloed channels if any channel is soloed
      if (solo) {
        this.channels.forEach((ch) => {
          if (ch.id !== id && ch.type !== 'master') {
            ch.isMuted = true
          }
        })
      } else if (!this.hasSoloedChannels()) {
        // Unmute all channels if no channels are soloed
        this.channels.forEach((ch) => {
          if (ch.type !== 'master') {
            ch.isMuted = false
          }
        })
      }

      this.recordAutomation(`${id}.solo`, solo ? 1 : 0)
    }
  }

  /**
   * Check if any channels are soloed
   */
  private hasSoloedChannels(): boolean {
    return Array.from(this.channels.values()).some(ch => ch.isSolo)
  }

  // ========== Effects ==========

  /**
   * Add insert effect to channel
   */
  addInsert(channelId: string, insert: InsertSlot): void {
    const channel = this.getChannel(channelId)
    if (channel) {
      channel.inserts.push(insert)
    }
  }

  /**
   * Remove insert effect from channel
   */
  removeInsert(channelId: string, insertId: string): void {
    const channel = this.getChannel(channelId)
    if (channel) {
      channel.inserts = channel.inserts.filter(insert => insert.id !== insertId)
    }
  }

  /**
   * Toggle insert bypass
   */
  toggleInsert(channelId: string, insertId: string): void {
    const channel = this.getChannel(channelId)
    if (channel) {
      const insert = channel.inserts.find(i => i.id === insertId)
      if (insert) {
        insert.enabled = !insert.enabled
      }
    }
  }

  /**
   * Add send effect to channel
   */
  addSend(channelId: string, send: Send): void {
    const channel = this.getChannel(channelId)
    if (channel) {
      channel.sends.push(send)
    }
  }

  /**
   * Remove send effect from channel
   */
  removeSend(channelId: string, sendId: string): void {
    const channel = this.getChannel(channelId)
    if (channel) {
      channel.sends = channel.sends.filter(send => send.id !== sendId)
    }
  }

  /**
   * Set send amount
   */
  setSendAmount(channelId: string, sendId: string, amount: number): void {
    const channel = this.getChannel(channelId)
    if (channel) {
      const send = channel.sends.find(s => s.id === sendId)
      if (send && amount >= 0 && amount <= 1) {
        send.amount = amount
        this.recordAutomation(`${channelId}.send.${sendId}`, amount)
      }
    }
  }

  // ========== Routing ==========

  /**
   * Set channel output bus
   */
  setOutputBus(id: string, bus: string): void {
    const channel = this.getChannel(id)
    if (channel && channel.type !== 'master') {
      channel.outputBus = bus
    }
  }

  // ========== Metering ==========

  /**
   * Update channel levels from audio backend
   */
  updateLevels(levels: Map<string, MeterData>): void {
    levels.forEach((levelData, channelId) => {
      const channel = this.getChannel(channelId)
      if (channel) {
        channel.levelL = levelData.levelL
        channel.levelR = levelData.levelR
        channel.peakL = levelData.peakL
        channel.peakR = levelData.peakR
      }
    })
  }

  /**
   * Get meter data for all channels
   */
  getAllMeterData(): Map<string, MeterData> {
    const meterData = new Map<string, MeterData>()

    this.channels.forEach((channel, id) => {
      meterData.set(id, {
        levelL: channel.levelL,
        levelR: channel.levelR,
        peakL: channel.peakL,
        peakR: channel.peakR,
        rmsL: -60, // TODO: Calculate from audio backend
        rmsR: -60
      })
    })

    // Add master
    meterData.set('master', {
      levelL: this.masterBus.levelL,
      levelR: this.masterBus.levelR,
      peakL: this.masterBus.peakL,
      peakR: this.masterBus.peakR,
      rmsL: -60,
      rmsR: -60
    })

    return meterData
  }

  // ========== Automation ==========

  /**
   * Record automation point
   */
  private recordAutomation(parameterId: string, value: number): void {
    if (!this.isPlayingAutomation) {
      let curve = this.automation.get(parameterId)
      if (!curve) {
        curve = {
          parameterId,
          points: []
        }
        this.automation.set(parameterId, curve)
      }

      curve.points.push({
        time: Date.now() / 1000, // Convert to seconds
        value,
        curve: 'linear'
      })
    }
  }

  /**
   * Get automation data for all parameters
   */
  getAutomationData(): AutomationCurve[] {
    return Array.from(this.automation.values())
  }

  /**
   * Get automation for specific parameter
   */
  getParameterAutomation(parameterId: string): AutomationCurve | undefined {
    return this.automation.get(parameterId)
  }

  /**
   * Clear all automation data
   */
  clearAutomation(): void {
    this.automation.clear()
  }

  // ========== Serialization ==========

  /**
   * Serialize console to JSON
   */
  toJSON(): object {
    return {
      channels: Array.from(this.channels.entries()),
      masterBus: this.masterBus,
      automation: Array.from(this.automation.entries())
    }
  }

  /**
   * Deserialize console from JSON
   */
  static fromJSON(json: any): MixingConsole {
    const console = new MixingConsole()

    // Restore channels
    if (json.channels) {
      json.channels.forEach(([id, channel]: [string, ChannelStrip]) => {
        console.channels.set(id, channel)
      })
    }

    // Restore master bus
    if (json.masterBus) {
      console.masterBus = json.masterBus
    }

    // Restore automation
    if (json.automation) {
      json.automation.forEach(([id, curve]: [string, AutomationCurve]) => {
        console.automation.set(id, curve)
      })
    }

    return console
  }
}
